lib/liquid/block.rb

```diff
@@ -87,7 +93,12 @@ def parse_body(body, tokens)
         parse_context.depth -= 1
       end
 
-      true
+      unless partial
+        body.remove_blank_strings if body.blank?
```

Comment:
Oh right, we can't really do the freezing in `parse_body` because we don't know whether to remove blank strings at this point.  We should only do this if the whole block tag is blank, which we don't really know until we finish parsing the whole tag.  I think that means that we can't really freeze the block bodies until after parsing the whole tag either, at least for these control flow tags.